dialog.1
'\" t
.\" $Id: dialog.1,v 1.240 2025/08/17 20:52:37 tom Exp $
.\" Copyright 2005-2024,2025  Thomas E. Dickey
.\"
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU Lesser General Public License, version 2.1
.\" as published by the Free Software Foundation.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" Lesser General Public License for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public
.\" License along with this program; if not, write to
.\"	Free Software Foundation, Inc.
.\"	51 Franklin St., Fifth Floor
.\"	Boston, MA 02110, USA.
.\"
.\" definitions for renaming
.ds p dialog
.ds l dialog
.ds L Dialog
.ds D DIALOG
.\"
.ie n .ds CW R
.el   \{
.ie \n(.g .ds CW CR
.el       .ds CW CW
.\}
.
.de ES
.ne 8
.IP
..
.de Ex
.RS +7
.PP
.nf
.ft \*(CW
..
.de Ee
.fi
.ft R
.RE
..
.\" Bulleted paragraph
.de bP
.ie n  .IP \(bu 4
.el    .IP \(bu 2
..
.ie \n(.g \{\
.ds `` \(lq
.ds '' \(rq
.ds '  \(aq
.\}
.el \{\
.ie t .ds `` ``
.el   .ds `` ""
.ie t .ds '' ''
.el   .ds '' ""
.ie t .ds '  \(aq
.el   .ds '  '
.\}",
.
.TH \*D "" 2025-08-17 "" "User commands"
.SH NAME
dialog \-
display dialog boxes from shell scripts
.SH SYNOPSIS
\fB\*p \-\-clear\fP
.br
.BI "\*p \-\-create\-rc " file
.br
\fB\*p \-\-print\-maxsize\fP
.br
\fB\*p\fP
\fIcommon-options\fP
\fIbox-options\fP
.SH DESCRIPTION
\fB\*L\fP
is a program that will let you present a variety of questions or
display messages using dialog boxes from a shell script.
These types of dialog boxes are implemented
(though not all are necessarily compiled into \fB\*p\fR):
.RS
.LP
.nh
.na
.BR buildlist ", "
.BR calendar ", "
.BR checklist ", "
.BR dselect ", "
.BR editbox ", "
.BR form ", "
.BR fselect ", "
.BR gauge ", "
.BR infobox ", "
.BR inputbox ", "
.BR inputmenu ", "
.BR menu ", "
.BR mixedform ", "
.BR mixedgauge ", "
.BR msgbox " (message), "
.BR passwordbox ", "
.BR passwordform ", "
.BR pause ", "
.BR prgbox ", "
.BR programbox ", "
.BR progressbox ", "
.BR radiolist ", "
.BR rangebox ", "
.BR tailbox ", "
.BR tailboxbg ", "
.BR textbox ", "
.BR timebox ", "
.BR treeview ", and "
.BR yesno " (yes/no)."
.ad
.RE
.PP
You can put more than one dialog box into a script:
.bP
Use the "\fB\-\-and\-widget\fP" token to force \fB\*p\fP to proceed to the next
dialog unless you have pressed ESC to cancel, or
.bP
Simply add the tokens for the next dialog box, making a chain.
\*L stops chaining when the return code from a dialog is nonzero,
e.g., Cancel or No (see DIAGNOSTICS).
.PP
Some widgets, e.g., checklist, will write text to \fB\*p\fP's output.
Normally that is the standard error, but there are options for
changing this:
\*(``\fB\-\-output\-fd\fP\*('',
\*(``\fB\-\-stderr\fP\*('' and
\*(``\fB\-\-stdout\fP\*(''.
No text is written if the Cancel button (or ESC) is pressed;
\fB\*p\fP exits immediately in that case.
.
.\" ************************************************************************
.SH OPTIONS
All options begin with \*(``\fB\-\-\fP\*(''
(two ASCII hyphens,
for the benefit of those using systems with deranged locale support).
.PP
A \*(``\fB\-\-\fP\*('' by itself is used as an escape,
i.e., the next token on the command-line is not treated as an option.
This is different from \fBgetopt\fP(1),
which uses that token to treat the remaining tokens as \fIparameters\fP
rather than \fIoptions\fP.
.RS
.PP
.B \*p \-\-title \-\- \-\-Not an option
.br
.B \*p \-\-title This \-\- \-\-title is not an option
.RE
.PP
\fB\*L\fP
uses no \fIparameters\fP,
and uses its own \fIoptions\fP parser.
.PP
When a common (e.g., non-widget) option is repeated,
the last found is the one that is used.
Boolean options are handled specially so they can be cancelled,
by adding (or omitting) a \*(``no\*('' modifier
after the leading \*(``\fB\-\-\fP\*(''.
For instance, \fB\-\-no\-shadow\fP is documented here,
but \fB\-\-shadow\fP also is accepted.
.PP
The \*(``\fB\-\-args\fP\*('' option tells \fB\*p\fP to list the command-line
parameters to the standard error.
This is useful when debugging complex scripts using
the \*(``\fB\-\-\fP\*('' and \*(``\fB\-\-file\fP\*('',
since the command-line may be rewritten as these are expanded.
.PP
The \*(``\fB\-\-file\fP\*('' option tells \fB\*p\fP to read parameters from
the file named as its value.
.RS
.B \*p \-\-file \fIparameterfile
.RE
.PP
Blanks not within double-quotes are discarded
(use backslashes to quote single characters).
The result is inserted into the command-line,
replacing \*(``\fB\-\-file\fP\*('' and its option value.
Interpretation of the command-line resumes from that point.
If \fIparameterfile\fP begins with \*(``&\*('', \fB\*p\fP
interprets the following text as a file descriptor number
rather than a filename.
.PP
Most widgets accept \fIheight\fP and \fIwidth\fP parameters,
which can be used to automatically size the widget to accommodate
multi-line message \fIprompt\fP values:
.bP
If the parameter is negative,
\fB\*l\fP uses the screen's size.
.bP
If the parameter is zero,
\fB\*l\fP uses minimum size for the widget to display the \fIprompt\fP
and data.
.bP
Otherwise, \fB\*l\fP uses the given size for the widget.
.
.SS "\fBCommon Options\fP"
Most of the common options are reset before processing each widget.
.
.IP "\fB\-\-ascii\-lines"
Rather than draw graphics lines around boxes,
draw ASCII \*(``+\*('' and \*(``-\*('' in the same place.
See also \*(``\fB\-\-no\-lines\fR\*(''.
.
.IP "\fB\-\-aspect \fIratio"
This gives you some control over the box dimensions when using auto
sizing (specifying 0 for height and width).
It represents width / height.
The default is 9, which means 9 characters wide to every 1 line high.
.
.IP "\fB\-\-backtitle \fIbacktitle"
Specifies a
\fIbacktitle\fP
string to be displayed on the backdrop, at the top of the screen.
.
.IP "\fB\-\-begin \fIy x"
Specify the position of the upper left corner of a dialog box on the screen.
.
.IP "\fB\-\-cancel\-label \fIstring"
Override the label used for \*(``Cancel\*('' buttons.
.
.IP "\fB\-\-clear"
Clears the widget screen, keeping only the screen_color background.
Use this when you combine widgets
with \*(``\fB\-\-and\-widget\fR\*('' to erase the
contents of a previous widget on the screen, so it won't be seen
under the contents of a following widget.
Understand this as the complement of \*(``\fB\-\-keep\-window\fR\*(''.
To compare the effects, use these:
.
.ES
All three widgets visible, staircase effect, ordered 1,2,3:
.Ex
\*p \e
                               \-\-begin 2 2 \-\-yesno "" 0 0 \e
    \-\-and\-widget               \-\-begin 4 4 \-\-yesno "" 0 0 \e
    \-\-and\-widget               \-\-begin 6 6 \-\-yesno "" 0 0
.Ee
.
.ES
Only the last widget is left visible:
.Ex
\*p \e
                 \-\-clear       \-\-begin 2 2 \-\-yesno "" 0 0 \e
    \-\-and\-widget \-\-clear       \-\-begin 4 4 \-\-yesno "" 0 0 \e
    \-\-and\-widget               \-\-begin 6 6 \-\-yesno "" 0 0
.Ee
.
.ES
All three widgets visible, staircase effect, ordered 3,2,1:
.Ex
\*p \e
                 \-\-keep\-window \-\-begin 2 2 \-\-yesno "" 0 0 \e
    \-\-and\-widget \-\-keep\-window \-\-begin 4 4 \-\-yesno "" 0 0 \e
    \-\-and\-widget               \-\-begin 6 6 \-\-yesno "" 0 0
.Ee
.
.ES
First and third widget visible, staircase effect, ordered 3,1:
.Ex
\*p \e
                 \-\-keep\-window \-\-begin 2 2 \-\-yesno "" 0 0 \e
    \-\-and\-widget \-\-clear       \-\-begin 4 4 \-\-yesno "" 0 0 \e
    \-\-and\-widget               \-\-begin 6 6 \-\-yesno "" 0 0
.Ee
.IP
Note, if you want to restore original console colors and send your
cursor home after the dialog program has exited, use the \fBclear\fR(1)
command.
Conversely, if you want to clear the screen and send your cursor to
the lower left after the \fB\*p\fP program has exited, use the
\fB\-\-erase\-on\-exit\fR\ option.
.
.IP "\fB\-\-color-mode \fIcode"
Extend the \fB\-\-colors\fP option to the content of programbox, tailbox
and textbox.
Use code \*(``2\*('' for this feature.
.
.IP "\fB\-\-colors"
Interpret embedded \*(``\eZ\*('' sequences in the dialog text
by the following character,
which tells \fB\*p\fP to set colors or video attributes:
.RS
.bP
0 through 7 are the ANSI color numbers used in curses:
black,
red,
green,
yellow,
blue,
magenta,
cyan and
white respectively.
.bP
Bold is set by 'b', reset by 'B'.
.bP
Reverse is set by 'r', reset by 'R'.
.bP
Underline is set by 'u', reset by 'U'.
.bP
The settings are cumulative, e.g., \*(``\eZb\eZ1\*('' makes the following text
bold (perhaps bright) red.
.bP
Restore normal settings with \*(``\eZn\*(''.
.RE
.
.IP "\fB\-\-column\-separator \fIstring"
Tell \fB\*p\fP to split data for radio/checkboxes and menus on the
occurrences of the given string, and to align the split data into columns.
.
.IP "\fB\-\-cr\-wrap"
Interpret embedded newlines in the dialog text as a newline on the screen.
Otherwise, \fB\*p\fR will only wrap lines
where needed to fit inside the text box.
.IP
Even though you can control line breaks with this,
\fB\*L\fR will still wrap any lines that are too long for the width of the box.
Without cr-wrap, the layout of your text may be formatted to look nice
in the source code of your script without affecting the way it will
look in the dialog.
.IP
The \fIcr\-wrap\fP feature is implemented subject to these conditions:
.RS
.bP
the string contains \*(``\en\*('' and the \fB\-\-no\-nl\-expand\fP option is
not used, or
.bP
the \fB\-\-trim\fP option is used.
.RE
.IP
For more information, see \fBWhitespace Options\fP.
.
.IP "\fB\-\-create\-rc \fIfile"
When
\fB\*p\fP
supports run-time configuration,
this can be used to dump a sample configuration file to the file specified
by
.IR file "."
.
.IP "\fB\-\-cursor\-off\-label"
Place the terminal cursor at the end of a button instead of on the
first character of the button label.
This is useful to reduce visual confusion when the cursor coloration
interacts poorly with the button-label text colors.
.
.IP "\fB\-\-date\-format \fIformat"
If the host provides \fBstrftime\fP,
this option allows you to specify the format of the date printed for
the \fB\-\-calendar\fP widget.
The time of day (hour, minute, second) are the current local time.
.
.IP "\fB\-\-defaultno"
Make the default value of the
\fByes/no\fP
box a
.BR No .
Likewise, treat the default button of widgets that provide
\*(``OK\*('' and \*(``Cancel\*(''
as a \fICancel\fP.
If \*(``\fB\-\-no\-cancel\fP\*('' or \*(``\fB\-\-visit\-items\fP\*('' are given
those options overrides this,
making the default button always \*(``Yes\*(''
(internally the same as \*(``OK\*('').
.
.IP "\fB\-\-default\-button \fIstring"
Set the default (preselected) button in a widget.
By preselecting a button,
a script makes it possible for the user to simply press \fIEnter\fP
to proceed through a dialog with minimum interaction.
.IP
The option's value is the name of the button:
.IR ok ,
.IR yes ,
.IR cancel ,
.IR no ,
.IR help "\ or"
.IR extra .
.IP
Normally the first button in each widget is the default.
The first button shown is determined by the widget
together with the \*(``\fB\-\-no\-ok\fP\*(''
and \*(``\fB\-\-no\-cancel\fP\*('' options.
If this option is not given, there is no default button assigned.
.
.IP "\fB\-\-default\-item \fIstring"
Set the default item in a checklist, form or menu box.
Normally the first item in the box is the default.
.
.IP "\fB\-\-erase\-on\-exit"
When \fB\*p\fP exits, remove the dialog widget, erasing the entire
screen to its native background color, and place the terminal cursor
at the lower left corner.
.
.IP "\fB\-\-exit\-label \fIstring"
Override the label used for \*(``EXIT\*('' buttons.
.
.IP "\fB\-\-extra\-button"
Show an extra button, between \*(``OK\*('' and \*(``Cancel\*('' buttons.
.IP
The extra button appears between \*(``Yes\*('' and \*(``No\*(''
for the \fIyesno\fP widget.
.
.IP "\fB\-\-extra\-label \fIstring"
Override the label used for \*(``Extra\*('' buttons.
Note: for inputmenu widgets, this defaults to \*(``Rename\*(''.
.
.IP "\fB\-\-help"
Prints the help message to the standard output and exits.
The help message is also printed if no options are given,
or if an unrecognized option is given.
.
.IP "\fB\-\-help\-button"
Show a help-button after \*(``OK\*('' and \*(``Cancel\*('' buttons
in boxes which have a list of tagged items
(i.e., checklist, radiolist, menu, and treeview boxes).
.IP
The help-button appears after \*(``Yes\*('' and \*(``No\*(''
for the \fIyesno\fP widget.
.IP
On exit, the return status indicates that the Help button was pressed.
\fB\*L\fP also writes a message to its output
after the token \*(``HELP\*('':
.RS
.bP

CHANGES
-- $Id: CHANGES,v 1.832 2025/08/17 20:08:23 tom Exp $
-- Thomas E. Dickey <dickey@invisible-island.net>

This version of dialog was originally from a Debian snapshot.  I've done this
to it:

2025/08/17
	+ amend macros for temporarily disabling --no-tags in buildlist and
	  treeview to allow --no-items (report/testcase by Maksym Telychko).
	+ configure script fixes for compiler warnings

2025/01/16
	+ modify dialog-config.in to allow for include-subdirectory.
	+ update configure script macros for changes made with Cdk.

2025/01/14
	+ use $ LIB_ROOTNAME, e.g., set by --with-libname, in generating
	  *-config and ".pc" files

2025/01/12
	+ further improvements to matching output from *-config and ".pc"
	+ improve gen-pkgconfig.in to mark the screen library, e.g., ncurses,
	  as "Required" and filter the cflags/libs from required packages out
	  of the ".pc" file, because pkg-config will add those automatically.

2025/01/09
	+ fix some gcc15 strict compiler warnings
	+ configure script fixes for compiler warnings
	+ update config.guess, config.sub

2024/06/19
	+ modify configure script to ease deprecation of libtool-bin package
	  (Debian #1071581).
	+ update dialog.map, dialog.sym for 2023/10/02 changes.
	+ improve dialog-config.in,
	  + suppressing several special cases of linker options, and
	  + adding support for --cflags-only-I, --cflags-only-other

2024/04/08
	+ fix a few cppcheck warnings
	+ improvements to NLS configuration.

2024/03/12
	+ add configure check for PIC vs PIE, needed for linking to libraries
	  which lack pkg-config scripts to specify LDFLAGS.

2024/03/07
	+ add option --color-modes, which can be used to color the content of
	  programbox, tailbox, textbox (requested by Rafał Radziejewski).

2024/03/06
	+ updated configure script, e.g., for compiler-warning fixes.
	+ amend change to formbox while revising --max-input to work with the
	  form's "ilen" parameter (report by Anna-Maria Gruber, cf: 2022/04/14)
	+ update config.guess, config.sub

2024/01/01
	+ updated configure script, e.g., for compiler-warning fixes.
	+ updated lv.po from
		http://translationproject.org/latest/dialog/

2023/10/02
	+ add/use dlg_print_nowrap(), to handle multibyte character strings in
	  progressbox and tailbox (report/testcase by Sergey Merzlikin).
	+ updated configure script, e.g., for compiler-warning fixes.
	+ update config.guess, config.sub

2023/08/18
	+ updated configure script, e.g., for compiler-warning fixes.
	+ minor fixes for manpages to address mandoc warnings.
	+ updated th.po from
		http://translationproject.org/latest/dialog/
	+ update config.guess, config.sub

2023/02/09
	+ install ".pc" files in the test-packages.
	+ add pkg-config ".pc" file, for use with the development library.

2023/02/07
	+ change shared-library configuration in test-packages to use the
	  ABI version in the library filename to match its SONAME.
	+ add sample scripts to test-packages, using new makefile rule
	  "install-examples".
	+ corrected cdialog*.install helper scripts in test-package for Debian
	+ change ftp URLs in test-packages to https
	+ updated configure script, e.g., for compiler-warning fixes.
	+ updated fur.po, zh_TW.po from
		http://translationproject.org/latest/dialog/

2022/12/29
	+ modify slackware.rc to match dialog's output from "make check" using
	  this configuration file (FreeBSD #267777).
	+ fix shellcheck warnings in helper-scripts for building/installing.
	+ updated configure script, e.g., for compiler-warning fixes.
	+ updated hu.po from
		http://translationproject.org/latest/dialog/
	+ update config.guess, config.sub, install-sh

2022/07/28
	+ initialize color-settings before --create-rc option is processed.
	+ add "use_scrollbar" setting in rc-file to use in whiptail.rc
	+ improve manual page discussion of "--" token.
	+ updated configure macro for math-library functions.
	+ updated ca.po from
		http://translationproject.org/latest/dialog/
	+ update config.guess

2022/06/15
	+ fix links in manpages (FreeBSD #264660).

2022/05/26
	+ document the built-in key bindings which are not part of widgets.
	+ updated da.po fa.po is.po, added ka.po from
		http://translationproject.org/latest/dialog/
	+ update config.sub

2022/04/14
	+ adjust fallback bindings for backspace/delete to match the curses
	  library's sense of backspace (Arch #74379).
	+ amend limit-checks to make --max-input work for sizes greater than
	  2048 (Debian #1004874).

2022/04/04
	+ corrected limit-checks in checklist.c, menubox.c, and treeview.c when
	  resizing while the data is scrolled (Debian #1004868).
	+ improve logic for updating form-height after screen-resizing.
	+ move most of the non-API definitions from dialog.h to dlg_internals.h
	+ updated configure macros.
	+ update config.guess, config.sub
	+ updated es.po, tr.po from
		http://translationproject.org/latest/dialog/

2022/01/17
	+ update test-package for Debian to use debhelper.
	+ adjust character-offset in editbox.c in case it is past the end of
	  the string after trimming (Debian #1003185).
	+ updated ro.po from
		http://translationproject.org/latest/dialog/

2021/12/14
	+ fix some inconsistent #if vs #ifdef.
	+ correct off-by-one limit for month when validating parameters of the
	  calendar widget (patch by "viverna", Debian #1000720).
	+ modify dlg_yes_labels to not switch to ok/cancel labels if the
	  extra button is used (report by Стрижак Андрей).
	+ move leak-checking definitions from dialog.h to dlg_internals.h

2021/11/07
	+ use mktemp utility rather than tempfile for temporary filename in
	  the sample scripts (patch by Ville Skyttä).
	+ updated id.po, lt.po from
		http://translationproject.org/latest/dialog/
	+ update config.guess, config.sub

2021/06/21
	+ directly include stddef.h for systems which need it for offsetof().
	+ modify gauge's handle_input function to use the file pointer in the
	  object in preference to the global dialog_state.
	+ correct height-computation for fselect and timebox, broken by fixes

dialog.pl
# Functions that handle calling dialog(1) -*-perl-*-
# $Id: dialog.pl,v 1.17 2018/06/12 01:27:04 tom Exp $
#
# The "rhs_" functions, as well as return_output originally came from Redhat
# 4.0, e.g.,
# http://www.ibiblio.org/pub/historic-linux/distributions/redhat-4.0/i386/live/usr/bin/Xconfigurator.pl
# The other functions were added to make this more useful for demonstrations.

# These comments are from the original file:
#------------------------------------------------------------------------------
# Return values are 1 for success and 0 for failure (or cancel)
# Resultant text (if any) is in dialog_result

# Unfortunately, the gauge requires use of /bin/sh to get going.
# I didn't bother to make the others shell-free, although it
# would be simple to do.

# Note that dialog generally returns 0 for success, so I invert the
# sense of the return code for more readable boolean expressions.
#------------------------------------------------------------------------------

use warnings;
use strict;
use diagnostics;

our $DIALOG = "dialog";
our $GAUGE;
our $gauge_width;
our $scr_lines = 24;
our $scr_cols  = 80;
our @dialog_result;
our $trace = 0;

require "flush.pl";

sub trace {
    if ($trace) {
        if ( open TRACE, ">>dialog.log" ) {
            printf TRACE $_[0], @_[ 1 .. $#_ ];
            close TRACE;
        }
    }
}

sub quoted($) {
    my $text = shift;
    $text =~ s/[\r\n]+/\n/g;
    $text =~ s/[^\n\t -~]/?/g;
    $text =~ s/([\\"])/\\$1/g;
    return sprintf "\"%s\"", $text;
}

sub screensize() {
    my $params = `$DIALOG --stdout --print-maxsize`;
    $params =~ s/\s+$//;
    $params =~ s/^[^:]*:\s+//;
    my @params = split /,\s+/, $params;
    if ( $#params == 1 ) {
        $scr_lines = $params[0];
        $scr_cols  = $params[1];
    }
    else {
        $scr_lines = 24;
        $scr_cols  = 80;
    }
}

sub height_of($$) {
    my $width   = shift;
    my $message = shift;
    my $command =
        "$DIALOG --stdout --print-text-size "
      . &quoted($message)
      . " $scr_lines $width 2>&1";
    my $params = `$command`;
    my @params = split( /\s/, $params );
    return $params[0];
}

sub rhs_clear {
    return system("$DIALOG --clear");
}

sub rhs_textbox {
    my ( $title, $file, $width, $height ) = @_;

    $width  = int($width);
    $height = int($height);
    system( "$DIALOG --title "
          . &quoted($title)
          . " --textbox $file $height $width" );

    return 1;
}

sub rhs_msgbox {
    my ( $title, $message, $width ) = @_;
    my ( $tmp, $height );

    $width   = int($width);
    $message = &rhs_wordwrap( $message, $width );
    $height  = 5 + &height_of( $width, $message );

    $tmp =
      system( "$DIALOG --title "
          . &quoted($title)
          . " --msgbox "
          . &quoted($message)
          . " $height $width" );
    if ($tmp) {
        return 0;
    }
    else {
        return 1;
    }
}

sub rhs_infobox {
    my ( $title, $message, $width ) = @_;
    my ( $tmp, $height );

    $width   = int($width);
    $message = &rhs_wordwrap( $message, $width );
    $height  = 2 + &height_of( $width, $message );

    return
      system( "$DIALOG --title "
          . &quoted($title)
          . " --infobox "
          . &quoted($message)
          . " $height $width" );
}

sub rhs_yesno {
    my ( $title, $message, $width ) = @_;
    my ( $tmp, $height );

    $width   = int($width);
    $message = &rhs_wordwrap( $message, $width );
    $height  = 4 + &height_of( $width, $message );

    $tmp =
      system( "$DIALOG --title "
          . &quoted($title)
          . " --yesno "
          . &quoted($message)
          . " $height $width" );

    # Dumb: dialog returns 0 for "yes" and 1 for "no"
    if ( !$tmp ) {
        return 1;
    }
    else {
        return 0;
    }
}

sub rhs_gauge {
    my ( $title, $message, $width, $percent ) = @_;
    my ( $tmp, $height );

    $width       = int($width);
    $gauge_width = $width;

    $message = &rhs_wordwrap( $message, $width );
    $height = 5 + &height_of( $width, $message );

    open( $GAUGE,
            "|$DIALOG --title "
          . &quoted($title)
          . " --gauge "
          . &quoted($message)
          . " $height $width $percent" );
}

sub rhs_update_gauge {
    my ($percent) = @_;

    &printflush( $GAUGE, "$percent\n" );
}

sub rhs_update_gauge_and_message {
    my ( $message, $percent ) = @_;

    $message = &rhs_wordwrap( $message, $gauge_width );
    $message =~ s/\n/\\n/g;
    &printflush( $GAUGE, "XXX\n$percent\n$message\nXXX\n" );
}

sub rhs_stop_gauge {
    close $GAUGE;
}

sub rhs_inputbox {
    my ( $title, $message, $width, $instr ) = @_;
    my ( $tmp, $height );

    $width   = int($width);
    $message = &rhs_wordwrap( $message, $width );
    $height  = 7 + &height_of( $width, $message );

    return &return_output( 0,
            "$DIALOG --title "
          . &quoted($title)
          . " --inputbox "
          . &quoted($message)
          . " $height $width "
          . &quoted($instr) );
}

sub rhs_menu {
    my ( $title, $message, $width, $numitems ) = @_;
    my ( $i, $tmp, $ent, $height, $listheight, $menuheight, @list );

    $width    = int($width);
    $numitems = int($numitems);

    shift;
    shift;
    shift;
    shift;

    @list = ();
    for ( $i = 0 ; $i < $numitems ; $i++ ) {
        $ent         = shift;
        $list[@list] = &quoted($ent);
        $ent         = shift;
        $list[@list] = &quoted($ent);
    }

    $message = &rhs_wordwrap( $message, $width );
    $listheight = &height_of( $width, $message );
    $height = 6 + $listheight + $numitems;

    if ( $height <= $scr_lines ) {
        $menuheight = $numitems;
    }
    else {
        $height     = $scr_lines;
        $menuheight = $scr_lines - $listheight - 6;
    }

    return &return_output( 0,
            "$DIALOG --title "
          . &quoted($title)
          . " --menu "
          . &quoted($message)
          . " $height $width $menuheight @list" );
}

sub rhs_menul {
    my ( $title, $message, $width, $numitems ) = @_;
    my ( $i, $tmp, $ent, $height, $listheight, $menuheight, @list );

    $width    = int($width);
    $numitems = int($numitems);

    shift;
    shift;
    shift;
    shift;

    @list = ();
    for ( $i = 0 ; $i < $numitems ; $i++ ) {
        $ent         = shift;
        $list[@list] = &quoted($ent);
        $list[@list] = &quoted("");
    }

    $message = &rhs_wordwrap( $message, $width );
    $listheight = &height_of( $width, $message );
    $height = 6 + $listheight + $numitems;

    if ( $height <= $scr_lines ) {
        $menuheight = $numitems;
    }
    else {
        $height     = $scr_lines;
        $menuheight = $scr_lines - $listheight - 6;
    }

    return &return_output( 0,
            "$DIALOG --title "
          . &quoted($title)
          . " --menu "
          . &quoted($message)
          . " $height $width $menuheight @list" );
}

sub rhs_menua {
    my ( $title, $message, $width, %items ) = @_;
    my ( $tmp, $ent, $height, $listheight, $menuheight, @list );

    $width = int($width);
    @list  = ();
    foreach $ent ( sort keys(%items) ) {
        $list[@list] = &quoted($ent);
        $list[@list] = &quoted( $items{$ent} );
    }

    my $numitems = keys(%items);
    $message = &rhs_wordwrap( $message, $width );
    $listheight = &height_of( $width, $message );
    $height = 6 + $listheight + $numitems;

    if ( $height <= $scr_lines ) {
        $menuheight = $numitems;
    }
    else {
        $height     = $scr_lines;
        $menuheight = $scr_lines - $listheight - 6;
    }

    return &return_output( 0,
            "$DIALOG --title "
          . &quoted($title)
          . " --menu "
          . &quoted($message)
          . " $height $width $menuheight @list" );
}

sub rhs_checklist {
    my ( $title, $message, $width, $numitems ) = @_;
    my ( $i, $tmp, $ent, $height, $listheight, $menuheight, @list );

    $width    = int($width);
    $numitems = int($numitems);

    shift;
    shift;
    shift;
    shift;

    @list = ();
    for ( $i = 0 ; $i < $numitems ; $i++ ) {
        $ent         = shift;
        $list[@list] = &quoted($ent);
        $ent         = shift;
        $list[@list] = &quoted($ent);
        $ent         = shift;
        if ($ent) {
            $list[@list] = "ON";
        }
        else {
            $list[@list] = "OFF";
        }
    }

    $message = &rhs_wordwrap( $message, $width );
    $listheight = &height_of( $width, $message );
    $height = 6 + $listheight + $numitems;

    if ( $height <= $scr_lines ) {
        $menuheight = $numitems;
    }
    else {
        $height     = $scr_lines;
        $menuheight = $scr_lines - $listheight - 6;
    }

    return &return_output( "list",
            "$DIALOG --title "
          . &quoted($title)
          . " --separate-output --checklist "
          . &quoted($message)
          . " $height $width $menuheight @list" );
}

sub rhs_checklistl {
    my ( $title, $message, $width, $numitems ) = @_;
    my ( $i, $tmp, $ent, $height, $listheight, $menuheight, @list );

    $width    = int($width);
    $numitems = int($numitems);

    shift;
    shift;
    shift;
    shift;

    @list = ();
    for ( $i = 0 ; $i < $numitems ; $i++ ) {
        $ent         = shift;
        $list[@list] = &quoted($ent);
        $list[@list] = &quoted("");
        $list[@list] = "OFF";
    }

    $message = &rhs_wordwrap( $message, $width );
    $listheight = &height_of( $width, $message );
    $height = 6 + $listheight + $numitems;

    if ( $height <= $scr_lines ) {
        $menuheight = $numitems;
    }
    else {
        $height     = $scr_lines;
        $menuheight = $scr_lines - $listheight - 6;
    }
    return &return_output( "list",
            "$DIALOG --title "
          . &quoted($title)
          . " --separate-output --checklist "
          . &quoted($message)
          . " $height $width $menuheight @list" );
}

sub rhs_checklista {
    my ( $title, $message, $width, %items ) = @_;
    my ( $tmp, $ent, $height, $listheight, $menuheight, @list );

    shift;
    shift;
    shift;
    shift;

    @list = ();
    foreach $ent ( sort keys(%items) ) {
        $list[@list] = &quoted($ent);
        $list[@list] = &quoted( $items{$ent} );
        $list[@list] = "OFF";
    }

    my $numitems = keys(%items);
    $message = &rhs_wordwrap( $message, $width );
    $listheight = &height_of( $width, $message );
    $height = 6 + $listheight + $numitems;

    if ( $height <= $scr_lines ) {
        $menuheight = $numitems;
    }
    else {
        $height     = $scr_lines;
        $menuheight = $scr_lines - $listheight - 6;
    }

    return &return_output( "list",
            "$DIALOG --title "
          . &quoted($title)
          . " --separate-output --checklist "
          . &quoted($message)
          . " $height $width $menuheight @list" );
}

sub rhs_radiolist {
    my ( $title, $message, $width, $numitems ) = @_;
    my ( $i, $tmp, $ent, $height, $listheight, $menuheight, @list );

    $width    = int($width);
    $numitems = int($numitems);

    shift;
    shift;
    shift;
    shift;

    @list = ();
    for ( $i = 0 ; $i < $numitems ; $i++ ) {
        $ent         = shift;
        $list[@list] = &quoted($ent);
        $ent         = shift;
        $list[@list] = &quoted($ent);
        $ent         = shift;
        if ($ent) {
            $list[@list] = "ON";
        }
        else {
            $list[@list] = "OFF";
        }
    }

    $message = &rhs_wordwrap( $message, $width );
    $listheight = &height_of( $width, $message );
    $height = 6 + $listheight + $numitems;

    if ( $height <= $scr_lines ) {
        $menuheight = $numitems;
    }
    else {
        $height     = $scr_lines;
        $menuheight = $scr_lines - $listheight - 6;
    }

    return &return_output( 0,
            "$DIALOG --title "
          . &quoted($title)
          . " --radiolist "
          . &quoted($message)
          . " $height $width $menuheight @list" );
}

sub return_output {
    my ( $listp, $command ) = @_;
    my ($res) = 1;

    pipe( PARENT_READER, CHILD_WRITER );

    # We have to fork (as opposed to using "system") so that the parent
    # process can read from the pipe to avoid deadlock.
    my ($pid) = fork;
    if ( $pid == 0 ) {    # child
        close(PARENT_READER);
        open( STDERR, ">&CHILD_WRITER" );
        exec($command);
        die("no exec");
    }
    if ( $pid > 0 ) {     # parent
        close(CHILD_WRITER);
        if ($listp) {
            @dialog_result = ();
            while (<PARENT_READER>) {
                chop;
                $dialog_result[@dialog_result] = $_;
            }
        }
        else {
            @dialog_result = <PARENT_READER>;
        }
        close(PARENT_READER);
        waitpid( $pid, 0 );
        $res = $?;
    }

    # Again, dialog returns results backwards
    if ( !$res ) {
        return 1;
    }
    else {
        return 0;
    }
}

sub rhs_wordwrap {
    my ( $intext, $width ) = @_;
    my ( $outtext, $i, $j, @lines, $wrap, @words, $pos, $pad );

    &trace( "rhs_wordwrap\n\tintext:%s\n\twidth:%d\n", $intext, $width );
    &screensize;
    $width   = int($width);
    $outtext = "";
    $pad     = 3;             # leave 3 spaces around each line
    $pos     = $pad;          # current insert position
    $wrap    = 0;             # 1 if we have been auto wrapping
    my $insert_nl = 0;        # 1 if we just did an absolute
                              # and we should preface any new text
                              # with a new line
    @lines = split( /\n/, $intext );

    for ( $i = 0 ; $i <= $#lines ; $i++ ) {

        if ( $lines[$i] =~ /^>/ ) {
            $outtext .= "\n" if ($insert_nl);
            $outtext .= "\n" if ($wrap);
            $lines[$i] =~ /^>(.*)$/;
            $outtext .= $1;
            $insert_nl = 1;
            $wrap      = 0;
            $pos       = $pad;
        }
        else {
            $wrap = 1;
            @words = split( /\s+/, $lines[$i] );
            for ( $j = 0 ; $j <= $#words ; $j++ ) {
                if ($insert_nl) {
                    $outtext .= "\n";
                    $insert_nl = 0;
                }
                if ( ( length( $words[$j] ) + $pos ) > $width - $pad ) {
                    $outtext .= "\n";
                    $pos = $pad;
                }
                $outtext .= $words[$j] . " ";
                $pos += length( $words[$j] ) + 1;
            }
        }
    }

    &trace( "\touttext:%s\n", $outtext );
    return $outtext;
}

############
1;
